Wallet: remove db mode string

S3RK opened the pull request:
This is a [follow-up](https://github.com/bitcoin/bitcoin/pull/19077#discussion_r500261927) for #19077 

This PR simplifies DB interface by removing mode string from `WalletDatabase` and `WalletBatch`.

The mode string was used to determine two flags for the instantiation of db connection:
1) read-only flag. Never used on connection level. And on batch level Is only used within `BerkeleyDatabase::Rewrite` where it's replaced with bool flag.
2) create flag. Is not required as we always check `require_existing` & `require_create` flags in `MakeDatabase()` before creating actual database instance. So we can safely default to always creating database if it doesn't exist yet.

S3RK committed:
wallet: remove db mode string

We never need to open database in read-only mode as it's controlled
separately for every batch.

Also we can safely create database if it doesn't exist already
because require_existing option is verified in MakeDatabase
before creating a new WalletDatabase instance.

achow101 commented on src/wallet/bdb.cpp:
```diff
@@ -305,20 +305,20 @@ BerkeleyDatabase::~BerkeleyDatabase()
 }
 
-BerkeleyBatch::BerkeleyBatch(BerkeleyDatabase& database, const char* pszMode, bool fFlushOnCloseIn) : pdb(nullptr), activeTxn(nullptr), m_cursor(nullptr), m_database(database)
+BerkeleyBatch::BerkeleyBatch(BerkeleyDatabase& database, const bool fReadOnly, bool fFlushOnCloseIn) : pdb(nullptr), activeTxn(nullptr), m_cursor(nullptr), m_database(database)
```
nit: for new variables the naming convention is `snake_case`.

promag replied:
334eadec38afd0a34edb990322fda0825118ec76

👍 and drop `this->` below.

BTW, why not just drop member `fReadOnly` since it's always `false`?

S3RK replied:
Changed the name and dropped `this->`

> BTW, why not just drop member fReadOnly since it's always false?

I kept it, because it's set to `true` in `BerkeleyDatabase::Rewrite`.

promag replied:
Where?

achow101 replied:
https://github.com/bitcoin/bitcoin/pull/20130/files#diff-b06753f6e4621f5bcd6cd27fed4137e70a262596b08f7f3a1416ea21977c53d4R467

achow101 commented:
Concept ACK

laanwj commented:
> read-only flag. Never used on connection level.

*Should* we have a way to open databases read-only? It seems this would be useful in some cases, e.g. in the past I remember discussion about the bitcoin-wallet tool for purely informational commands, which should be able to leave a database unmodified otherwise (or be able to act on a read-only file system).

achow101 commented:
> _Should_ we have a way to open databases read-only? It seems this would be useful in some cases, e.g. in the past I remember discussion about the bitcoin-wallet tool for purely informational commands, which should be able to leave a database unmodified otherwise (or be able to act on a read-only file system).

I think that it would be useful to have a read-only mode in the future, but it should be implemented differently from this mode string that we currently have. So this PR is at least a first step towards that, and we can add read-only mode for when we need it.

promag reviewed (commented):
Tested ACK 53718b268a37d6501fbb65841e90a30a9c73e85f, small commits that could be squashed but no big deal.

S3RK commented:
> > read-only flag. Never used on connection level.
> 
> _Should_ we have a way to open databases read-only? It seems this would be useful in some cases, e.g. in the past I remember discussion about the bitcoin-wallet tool for purely informational commands, which should be able to leave a database unmodified otherwise (or be able to act on a read-only file system).

Though I agree we might want to have it I think it's better not to keep unused functionality in the code. Also, current implementation is pretty limited. We use one connection per wallet and this prevents us from opening a true read-only connection, because it will break some wallet functionality. I guess this is why read-only mode was never used in practice.

If we want to introduce proper read-only mode we need to discuss different design trade-offs. One option would be to look at the possibility to bring it to a wallet level (for example load wallet in read-only mode). Alternatively for a command-level read-only mode we need to: a) either manage it on a batch level b) create separate read-write and read-only connections c) reconnect when needed d) other ideas..

I'll be happy to continue the work if there is strong demand for the feature.

S3RK commented:
@promag thanks for the quick review! I fixed the comment and squashed the commits.

Side node: I don't know what is the accepted practice here, but I usually do small commits as it's easier to squash than split. I also believe it sometimes can be useful as it provides more information on implementation.

achow101 commented:
ACK 135afa749c6e835ea33b8678cdb35da9640eede8

laanwj commented:
> I think that it would be useful to have a read-only mode in the future, but it should be implemented differently from this mode string that we currently have. So this PR is at least a first step towards that, and we can add read-only mode for when we need it.

Fair enough, concept ACK then.
